### DISCUSSION TITLE
Committed instances expire after rollback

### DISCUSSION THREAD

**Discussion by ave-satan** [sqlalchemy/sqlalchemy#7587] [Usage Questions] [answered]
It looks like when rolling back any changes on a session it would expire everything, even though some changed instances were committed to the database and subsequent queries use new transactions.
That means when I try to access attributes on an expired instance it would issue a select query.

The problem is when using async alchemy it would crash on attribute access with that `MissingGreenlet` error, unless I manually call `session.refresh()` for the instance. I find it very confusing and annoying, because looking through that code I'm assuming my instances would stay there since they were committed on a separate transaction.

Does this look like a bug? Or is it intentional?

Python: 3.9.2
SQLAlchemy: 1.4.29

Check this sample for reference:
```python
import sqlalchemy as sa
import sqlalchemy.orm


engine = sa.create_engine(
    'postgresql://user:password@localhost/test',
    future=True,
    echo=True,
)

Base = sa.orm.declarative_base()


class Table(Base):
    __tablename__ = 'table'

    id = sa.Column(sa.Integer, primary_key=True)
    value = sa.Column(sa.String)

    def __repr__(self):
        return f'<Table({self.id}, {self.value})>'


print('Create table')
Base.metadata.create_all(engine)

print('Insert some data')
with engine.connect() as conn, conn.begin():
    conn.execute(
        sa.insert(Table).values(value='initial')
    )

session = sa.orm.Session(engine, expire_on_commit=False)

print('Load one row with session')
row = session.execute(sa.select(Table).where(Table.value == 'initial')).scalar()
print(row)

print('Update the row with session')
row.value = 'changed'
session.commit()

print('Insert new row with session')
session.add(Table(value='new'))
session.flush()

print('Rollback session')
session.rollback()

print('Access first row data')
print(row)

print('Update first row')
row.value = 'final'
session.commit()
```

![image](https://user-images.githubusercontent.com/9197588/150102133-16335987-c756-462d-a95a-a9660d841e81.png)

**Comment by zzzeek** [comment, marked as answer]
this is the expected behavior of the session.   you wouldn't be calling rollback() unless you are handling some error condition.  if you want to be able to rollback parts of the state of the transaction while leaving the rest alone, you would use begin_nested() to use a SAVEPOINT.  when the savepoint is rolled back, it only expires the objects that were actually affected by that sub-transaction, so that's probably what you are looking for here.
